File: test/scoped_broadcasts/mutate_broadcast_carrier.py
#!/usr/bin/env python3
import pathlib
import re
import sys


source = pathlib.Path(sys.argv[1])
target = pathlib.Path(sys.argv[2])
attack = sys.argv[3]
payload = source.read_bytes()


def mutate_match(pattern: bytes, label: str, group: int) -> None:
    global payload
    matches = list(re.finditer(pattern, payload))
    if not matches:
        raise SystemExit(f"{attack}: no {label} carrier found")
    match = matches[0]
    value = bytearray(match.group(group))
    candidates = [
        index
        for index, byte in enumerate(value)
        if chr(byte) in "0123456789abcdef"
    ]
    if not candidates:
        raise SystemExit(f"{attack}: {label} has no mutable identity byte")
    index = candidates[-1]
    value[index] = ord("0") if value[index] != ord("0") else ord("1")
    start, stop = match.span(group)
    payload = payload[:start] + bytes(value) + payload[stop:]


def mutate_decimal(pattern: bytes, label: str, group: int) -> None:
    global payload
    matches = list(re.finditer(pattern, payload))
    if not matches:
        raise SystemExit(f"{attack}: no {label} span found")
    match = matches[0]
    value = bytearray(match.group(group))
    index = len(value) - 1
    value[index] = ord("0") if value[index] != ord("0") else ord("1")
    start, stop = match.span(group)
    payload = payload[:start] + bytes(value) + payload[stop:]


if attack == "declaration-id":
    mutate_match(
        rb"(declaration\.[0-9a-f]{32})\|[0-9]+\|[0-9]+",
        "declaration",
        1,
    )
elif attack == "carrier-id":
    mutate_match(
        rb"structure\|(structure\.[0-9a-f]{32})\|\|[0-9]+\|[0-9]+",
        "structure",
        1,
    )
elif attack == "group-id":
    mutate_match(
        rb"group\|(group\.[0-9a-f]{32})\|[A-Za-z_][A-Za-z0-9_']*\|[0-9]+\|[0-9]+",
        "group",
        1,
    )
elif attack == "marker":
    old = b"verocaml:broadcast:carrier:v1:"
    new = b"verocaml:broadcast:carriem:v1:"
    if payload.count(old) == 0 or len(old) != len(new):
        raise SystemExit(f"{attack}: marker target is absent")
    payload = payload.replace(old, new, 1)
elif attack == "declaration-span":
    mutate_decimal(
        rb"declaration\.[0-9a-f]{32}\|([0-9]+)\|[0-9]+",
        "declaration",
        1,
    )
elif attack == "carrier-span":
    mutate_decimal(
        rb"structure\|structure\.[0-9a-f]{32}\|\|([0-9]+)\|[0-9]+",
        "structure",
        1,
    )
elif attack == "group-span":
    mutate_decimal(
        rb"group\|group\.[0-9a-f]{32}\|[A-Za-z_][A-Za-z0-9_']*\|([0-9]+)\|[0-9]+",
        "group",
        1,
    )
elif attack == "scope-span":
    mutate_decimal(
        rb"expression\.[0-9a-f]{32}\|([0-9]+)\|[0-9]+",
        "expression scope",
        1,
    )
elif attack in ("legacy-lemma", "legacy-axiom"):
    old = b"declaration."
    new = b"lemma|lemma." if attack == "legacy-lemma" else b"axiom|axiom."
    if payload.count(old) == 0 or len(old) != len(new):
        raise SystemExit(f"{attack}: neutral declaration carrier is absent")
    payload = payload.replace(old, new, 1)
else:
    raise SystemExit(f"unknown attack {attack}")

target.write_bytes(payload)

File: test/scoped_broadcasts/mutate_broadcast_carrier.ml
let fail format =
  Printf.ksprintf
    (fun message ->
      prerr_endline message;
      exit 1)
    format

let read_file path =
  let channel = open_in_bin path in
  Fun.protect
    ~finally:(fun () -> close_in channel)
    (fun () -> really_input_string channel (in_channel_length channel))

let write_file path payload =
  let channel = open_out_bin path in
  Fun.protect
    ~finally:(fun () -> close_out channel)
    (fun () -> output_bytes channel payload)

let has_literal payload offset literal =
  let literal_length = String.length literal in
  let rec loop index =
    if index = literal_length then true
    else if Bytes.get payload (offset + index) <> literal.[index] then false
    else loop (index + 1)
  in
  offset >= 0
  && offset + literal_length <= Bytes.length payload
  && loop 0

let find_literal ?(from = 0) payload literal =
  let last = Bytes.length payload - String.length literal in
  let rec loop offset =
    if offset > last then None
    else if has_literal payload offset literal then Some offset
    else loop (offset + 1)
  in
  loop from

let is_decimal = function '0' .. '9' -> true | _ -> false

let is_hexadecimal = function
  | '0' .. '9' | 'a' .. 'f' -> true
  | _ -> false

let is_identifier_start = function
  | 'A' .. 'Z' | 'a' .. 'z' | '_' -> true
  | _ -> false

let is_identifier_continue = function
  | 'A' .. 'Z' | 'a' .. 'z' | '0' .. '9' | '_' | '\'' -> true
  | _ -> false

let skip_while predicate payload offset =
  let rec loop index =
    if index < Bytes.length payload && predicate (Bytes.get payload index) then
      loop (index + 1)
    else index
  in
  loop offset

let has_hexadecimal payload offset length =
  let rec loop index =
    index = length
    ||
    (offset + index < Bytes.length payload
    && is_hexadecimal (Bytes.get payload (offset + index))
    && loop (index + 1))
  in
  loop 0

let decimal_before_bar payload offset =
  if offset >= Bytes.length payload || not (is_decimal (Bytes.get payload offset))
  then None
  else
    let stop = skip_while is_decimal payload offset in
    if stop < Bytes.length payload && Bytes.get payload stop = '|' then
      Some stop
    else None

type carrier = {
  identity_last : int;
  span_stop : int;
}

let find_declaration payload =
  let prefix = "declaration." in
  let rec loop from =
    match find_literal ~from payload prefix with
    | None -> None
    | Some start ->
        let hexadecimal = start + String.length prefix in
        let separator = hexadecimal + 32 in
        if
          has_hexadecimal payload hexadecimal 32
          && separator < Bytes.length payload
          && Bytes.get payload separator = '|'
        then
          let span_start = separator + 1 in
          (match decimal_before_bar payload span_start with
          | Some span_stop
            when span_stop + 1 < Bytes.length payload
                 && is_decimal (Bytes.get payload (span_stop + 1)) ->
              Some
                {
                  identity_last = separator - 1;
                  span_stop;
                }
          | _ -> loop (start + 1))
        else loop (start + 1)
  in
  loop 0

let find_structure payload =
  let prefix = "structure|structure." in
  let rec loop from =
    match find_literal ~from payload prefix with
    | None -> None
    | Some start ->
        let hexadecimal = start + String.length prefix in
        let separator = hexadecimal + 32 in
        if
          has_hexadecimal payload hexadecimal 32
          && separator + 1 < Bytes.length payload
          && Bytes.get payload separator = '|'
          && Bytes.get payload (separator + 1) = '|'
        then
          let span_start = separator + 2 in
          (match decimal_before_bar payload span_start with
          | Some span_stop
            when span_stop + 1 < Bytes.length payload
                 && is_decimal (Bytes.get payload (span_stop + 1)) ->
              Some
                {
                  identity_last = separator - 1;
                  span_stop;
                }
          | _ -> loop (start + 1))
        else loop (start + 1)
  in
  loop 0

let find_group payload =
  let prefix = "group|group." in
  let rec loop from =
    match find_literal ~from payload prefix with
    | None -> None
    | Some start ->
        let hexadecimal = start + String.length prefix in
        let separator = hexadecimal + 32 in
        if
          has_hexadecimal payload hexadecimal 32
          && separator + 1 < Bytes.length payload
          && Bytes.get payload separator = '|'
          && is_identifier_start (Bytes.get payload (separator + 1))
        then
          let name_stop =
            skip_while is_identifier_continue payload (separator + 2)
          in
          if
            name_stop < Bytes.length payload
            && Bytes.get payload name_stop = '|'
          then
            let span_start = name_stop + 1 in
            (match decimal_before_bar payload span_start with
            | Some span_stop
              when span_stop + 1 < Bytes.length payload
                   && is_decimal (Bytes.get payload (span_stop + 1)) ->
                Some
                  {
                    identity_last = separator - 1;
                    span_stop;
                  }
            | _ -> loop (start + 1))
          else loop (start + 1)
        else loop (start + 1)
  in
  loop 0

let find_expression payload =
  let prefix = "expression." in
  let rec loop from =
    match find_literal ~from payload prefix with
    | None -> None
    | Some start ->
        let hexadecimal = start + String.length prefix in
        let separator = hexadecimal + 32 in
        if
          has_hexadecimal payload hexadecimal 32
          && separator < Bytes.length payload
          && Bytes.get payload separator = '|'
        then
          let span_start = separator + 1 in
          (match decimal_before_bar payload span_start with
          | Some span_stop
            when span_stop + 1 < Bytes.length payload
                 && is_decimal (Bytes.get payload (span_stop + 1)) ->
              Some
                {
                  identity_last = separator - 1;
                  span_stop;
                }
          | _ -> loop (start + 1))
        else loop (start + 1)
  in
  loop 0

let toggle_zero payload index =
  Bytes.set payload index (if Bytes.get payload index = '0' then '1' else '0')

let mutate_identity attack label find payload =
  match find payload with
  | None -> fail "%s: no %s carrier found" attack label
  | Some carrier -> toggle_zero payload carrier.identity_last

let mutate_span attack label find payload =
  match find payload with
  | None -> fail "%s: no %s span found" attack label
  | Some carrier -> toggle_zero payload (carrier.span_stop - 1)

let replace_first attack payload old_value new_value absent_message =
  match find_literal payload old_value with
  | None -> fail "%s: %s" attack absent_message
  | Some start ->
      if String.length old_value <> String.length new_value then
        fail "%s: %s" attack absent_message;
      Bytes.blit_string new_value 0 payload start (String.length new_value)

let () =
  let source, target, attack =
    match Sys.argv with
    | [| _; source; target; attack |] -> (source, target, attack)
    | _ -> fail "usage: mutate_broadcast_carrier SOURCE TARGET ATTACK"
  in
  let payload = Bytes.of_string (read_file source) in
  (match attack with
  | "declaration-id" ->
      mutate_identity attack "declaration" find_declaration payload
  | "carrier-id" -> mutate_identity attack "structure" find_structure payload
  | "group-id" -> mutate_identity attack "group" find_group payload
  | "marker" ->
      replace_first attack payload "verocaml:broadcast:carrier:v1:"
        "verocaml:broadcast:carriem:v1:" "marker target is absent"
  | "declaration-span" ->
      mutate_span attack "declaration" find_declaration payload
  | "carrier-span" -> mutate_span attack "structure" find_structure payload
  | "group-span" -> mutate_span attack "group" find_group payload
  | "scope-span" ->
      mutate_span attack "expression scope" find_expression payload
  | "legacy-lemma" ->
      replace_first attack payload "declaration." "lemma|lemma."
        "neutral declaration carrier is absent"
  | "legacy-axiom" ->
      replace_first attack payload "declaration." "axiom|axiom."
        "neutral declaration carrier is absent"
  | _ -> fail "unknown attack %s" attack);
  write_file target payload
